x-pack/plugins/cases/public/components/all_cases/use_all_cases_state.tsx

```diff
@@ -5,244 +5,183 @@
-    DEFAULT_QUERY_PARAMS.sortField;
-
-  result.sortOrder =
-    params.sortOrder ??
-    urlParams.sortOrder ??
-    localStorageQueryParams?.sortOrder ??
-    DEFAULT_QUERY_PARAMS.sortOrder;
-
-  result.page = params.page ?? urlParams.page ?? DEFAULT_QUERY_PARAMS.page;
-
-  return result;
-};
+export function useAllCasesState(isModalView: boolean = false): UseAllCasesStateReturn {
+  const isStateLoadedFromLocalStorage = useRef(false);
+  const [tableState, setTableState] = useState<AllCasesTableState>(DEFAULT_CASES_TABLE_STATE);
+  const [urlState, setUrlState] = useAllCasesUrlState();
+  const [localStorageState, setLocalStorageState] = useAllCasesLocalStorage();
 
-const validateQueryParams = (queryParams: QueryParams): QueryParams => {
-  const perPage = Math.min(
-    queryParams.perPage,
-    CASES_TABLE_PERPAGE_VALUES[CASES_TABLE_PERPAGE_VALUES.length - 1]
+  const allCasesTableState: AllCasesTableState = useMemo(
+    () => (isModalView ? tableState : getAllCasesTableState(urlState, localStorageState)),
+    [isModalView, tableState, urlState, localStorageState]
   );
-  const sortOrder = !SORT_ORDER_VALUES.includes(queryParams.sortOrder)
-    ? DEFAULT_QUERY_PARAMS.sortOrder
-    : queryParams.sortOrder;
 
-  return { ...queryParams, perPage, sortOrder };
-};
+  const setState = useCallback(
+    (state: AllCasesTableState) => {
+      if (isModalView) {
+        setTableState(state);
+        return;
+      }
 
-/**
- * Previously, 'status' and 'severity' were represented as single options (strings).
- * To maintain backward compatibility while transitioning to the new type of string[],
- * we map the legacy type to the new type.
- */
-const convertToFilterOptionArray = (value: string | string[] | undefined) => {
-  if (typeof value === 'string') {
-    return [value];
-  }
-  return value;
-};
+      if (!deepEqual(state, urlState)) {
+        setUrlState(state);
+      }
 
-const getFilterOptions = (
-  filterOptions: FilterOptions,
-  params: FilterOptions,
-  urlParams: PartialFilterOptions,
-  localStorageFilterOptions?: PartialFilterOptions
-): FilterOptions => {
-  const severity =
-    params?.severity ??
-    urlParams?.severity ??
-    convertToFilterOptionArray(localStorageFilterOptions?.severity) ??
-    DEFAULT_FILTER_OPTIONS.severity;
-
-  const status =
-    params?.status ??
-    urlParams?.status ??
-    convertToFilterOptionArray(localStorageFilterOptions?.status) ??
-    DEFAULT_FILTER_OPTIONS.status;
+      if (!deepEqual(state, localStorageState)) {
+        setLocalStorageState(state);
+      }
+    },
+    [localStorageState, urlState, isModalView, setLocalStorageState, setUrlState]
+  );
+
+  // use of useEffect because setUrlState calls history.push
+  useEffect(() => {
+    if (
+      !isStateLoadedFromLocalStorage.current &&
+      isURLStateEmpty(urlState) &&
+      localStorageState &&
+      !deepEqual(urlState, localStorageState) &&
+      !isModalView
+    ) {
```

Comment:
One of the test scenarios is:

> Test that submitting a URL takes priority over the state persisted in the local storage.

This works as expected. 👍 

But defining a URL does not overwrite the local storage and I think it should.*

The local storage is only ever updated when a filter is clicked but what if I write a specific URL and go there? I would probably want it to be saved too.

In that scenario, the whole page is rendered but `set` is never called.

Maybe we could do:
```suggestion
    if (
      !isStateLoadedFromLocalStorage.current &&
      localStorageState &&
      !deepEqual(urlState, localStorageState) &&
      !isModalView
    ) {
```

* This ties in to [this comment](https://github.com/elastic/kibana/pull/175237#issuecomment-1934154403). If I have some "broken" field in the URL I can only remove it by deleting URL and local storage.

Reply:
Good point. Fixed in [`feae267` (#175237)](https://github.com/elastic/kibana/pull/175237/commits/feae26737b5a996c8229ebd64c6703050384108f).